File: October_6/solution.cpp
class TimeMap {
public:
    map<string, vector<pair<int,string>>> mp;
//     TimeMap() {
        
//     }
    
    void set(string key, string value, int timestamp) {
        mp[key].push_back({timestamp,value});
    }
    
    string get(string key, int timestamp) {
        string ans="";
        // for(auto &it : mp[key] )
        // {
        //     if(it.first>timestamp) break;
        //     ans=it.second;
        // }
        int left=0,right=mp[key].size()-1,mid;
        while(left<=right)
        {
            mid=(left+right)/2;
            if(mp[key][mid].first<timestamp) left=mid+1;
            else if(mp[key][mid].first==timestamp) return mp[key][mid].second;
            else right=mid-1;
            
        }
        if(right>=0) ans=mp[key][right].second;
        return ans;
    }
};

File: October_6/Solution.py
class TimeMap:

    def _init_(self):
        self.main=defaultdict(list)
        

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.main[key].append((value,timestamp))
        

    def get(self, key: str, timestamp: int) -> str:
        # print (self.main)
        if key in self.main:
            l,r=0,len(self.main[key])-1
            while l<r:
                mid=(l+r)//2
                if self.main[key][mid][1]>timestamp:
                    r=mid-1
                elif self.main[key][mid][1]<timestamp:
                    l=mid+1
                else:
                    return self.main[key][mid][0]
            if  self.main[key][l][1]<=timestamp:
                return self.main[key][l][0]
            elif self.main[key][l-1][1]<=timestamp:
                return self.main[key][l-1][0]
        return ""
